From f3c89b795ffbeff21034de8cdb13a51105b4f2fa Mon Sep 17 00:00:00 2001
From: David Disseldorp <ddiss@suse.de>
Date: Fri, 23 Dec 2016 11:37:53 +0100
Subject: [PATCH 02/11] target: add XCOPY target/segment desc sense codes
Patch-mainline: v4.10-rc4
References: bsc#991273
Git-commit: e864212078ded276bdb272b2e0ee6a979357ca8a

As defined in http://www.t10.org/lists/asc-num.htm. To be used during
validation of XCOPY target and segment descriptor lists.

Signed-off-by: David Disseldorp <ddiss@suse.de>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Bart Van Assche <bart.vanassche@sandisk.com>

[ddiss@suse.de: rebase against SLE12SP1 with sense switch]

---
 drivers/target/target_core_transport.c | 48 ++++++++++++++++++++++++++++++++++
 include/target/target_core_base.h      |  4 +++
 2 files changed, 52 insertions(+)

diff --git a/drivers/target/target_core_transport.c b/drivers/target/target_core_transport.c
index 2186180..fc4420d 100644
--- a/drivers/target/target_core_transport.c
+++ b/drivers/target/target_core_transport.c
@@ -1650,6 +1650,10 @@ void transport_generic_request_failure(struct se_cmd *cmd,
 	case TCM_LOGICAL_BLOCK_REF_TAG_CHECK_FAILED:
 	case TCM_MISCOMPARE_VERIFY:
 	case TCM_COPY_TARGET_DEVICE_NOT_REACHABLE:
+	case TCM_TOO_MANY_TARGET_DESCS:
+	case TCM_UNSUPPORTED_TARGET_DESC_TYPE_CODE:
+	case TCM_TOO_MANY_SEGMENT_DESCS:
+	case TCM_UNSUPPORTED_SEGMENT_DESC_TYPE_CODE:
 		break;
 	case TCM_OUT_OF_RESOURCES:
 		sense_reason = TCM_LOGICAL_UNIT_COMMUNICATION_FAILURE;
@@ -2814,6 +2818,50 @@ transport_send_check_condition_and_sense(struct se_cmd *cmd,
 		/* INVALID FIELD IN PARAMETER LIST */
 		buffer[SPC_ASC_KEY_OFFSET] = 0x26;
 		break;
+	case TCM_TOO_MANY_TARGET_DESCS:
+		/* CURRENT ERROR */
+		buffer[0] = 0x70;
+		buffer[SPC_ADD_SENSE_LEN_OFFSET] = 10;
+		/* ILLEGAL REQUEST */
+		buffer[SPC_SENSE_KEY_OFFSET] = ILLEGAL_REQUEST;
+		/* INVALID FIELD IN PARAMETER LIST */
+		buffer[SPC_ASC_KEY_OFFSET] = 0x26;
+		/* TOO MANY TARGET DESCRIPTORS */
+		buffer[SPC_ASCQ_KEY_OFFSET] = 0x06;
+		break;
+	case TCM_UNSUPPORTED_TARGET_DESC_TYPE_CODE:
+		/* CURRENT ERROR */
+		buffer[0] = 0x70;
+		buffer[SPC_ADD_SENSE_LEN_OFFSET] = 10;
+		/* ILLEGAL REQUEST */
+		buffer[SPC_SENSE_KEY_OFFSET] = ILLEGAL_REQUEST;
+		/* INVALID FIELD IN PARAMETER LIST */
+		buffer[SPC_ASC_KEY_OFFSET] = 0x26;
+		/* UNSUPPORTED TARGET DESCRIPTOR TYPE CODE */
+		buffer[SPC_ASCQ_KEY_OFFSET] = 0x07;
+		break;
+	case TCM_TOO_MANY_SEGMENT_DESCS:
+		/* CURRENT ERROR */
+		buffer[0] = 0x70;
+		buffer[SPC_ADD_SENSE_LEN_OFFSET] = 10;
+		/* ILLEGAL REQUEST */
+		buffer[SPC_SENSE_KEY_OFFSET] = ILLEGAL_REQUEST;
+		/* INVALID FIELD IN PARAMETER LIST */
+		buffer[SPC_ASC_KEY_OFFSET] = 0x26;
+		/* TOO MANY SEGMENT DESCRIPTORS */
+		buffer[SPC_ASCQ_KEY_OFFSET] = 0x08;
+		break;
+	case TCM_UNSUPPORTED_SEGMENT_DESC_TYPE_CODE:
+		/* CURRENT ERROR */
+		buffer[0] = 0x70;
+		buffer[SPC_ADD_SENSE_LEN_OFFSET] = 10;
+		/* ILLEGAL REQUEST */
+		buffer[SPC_SENSE_KEY_OFFSET] = ILLEGAL_REQUEST;
+		/* INVALID FIELD IN PARAMETER LIST */
+		buffer[SPC_ASC_KEY_OFFSET] = 0x26;
+		/* UNSUPPORTED SEGMENT DESCRIPTOR TYPE CODE */
+		buffer[SPC_ASCQ_KEY_OFFSET] = 0x09;
+		break;
 	case TCM_PARAMETER_LIST_LENGTH_ERROR:
 		/* CURRENT ERROR */
 		buffer[0] = 0x70;
diff --git a/include/target/target_core_base.h b/include/target/target_core_base.h
index 9a67ab9..bdf0b09 100644
--- a/include/target/target_core_base.h
+++ b/include/target/target_core_base.h
@@ -214,6 +214,10 @@ enum tcm_sense_reason_table {
 	TCM_LOGICAL_BLOCK_APP_TAG_CHECK_FAILED	= R(0x16),
 	TCM_LOGICAL_BLOCK_REF_TAG_CHECK_FAILED	= R(0x17),
 	TCM_COPY_TARGET_DEVICE_NOT_REACHABLE	= R(0x18),
+	TCM_TOO_MANY_TARGET_DESCS		= R(0x19),
+	TCM_UNSUPPORTED_TARGET_DESC_TYPE_CODE	= R(0x1a),
+	TCM_TOO_MANY_SEGMENT_DESCS		= R(0x1b),
+	TCM_UNSUPPORTED_SEGMENT_DESC_TYPE_CODE	= R(0x1c),
 #undef R
 };
 
-- 
2.10.2

